library/gcm.c

```diff
@@ -442,6 +464,11 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
         return( MBEDTLS_ERR_GCM_BAD_INPUT );
     }
 
+    if ( ( ctx->len == 0 ) && ( ctx->add_len % 16 ) )
+    {
```

Comment:
Partly on me when I worked on `update` but it's getting worse: please add comment in `gcm.c` (in the documentation of `mbedtls_gcm_context` would be better but it would conflict with #4511 so let's avoid that so let's put it at the top of the implementation of `mbedtls_gcm_update_ad` maybe?) explaining what `buf` contains at various stages:
* When `add_len % 16 == 0` and `len == 0`: initial state.
* When `add_len % 16 != 01 and `len == 0`: the first `add_len % 16` bytes of `buf` have a partial AD block xored in and not yet multiplied in.
* When `len != 0`: the first `add_len % 16` bytes of `buf` have partial ciphertext xored in and not yet multiplied in.